sdk/platform/win32/src/az_win32.c

```diff
@@ -1,34 +1,22 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
-#include <az_platform_internal.h>
+#include <az_platform.h>
+
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <windows.h>
 
 #include <_az_cfg.h>
 
 AZ_NODISCARD int64_t az_platform_clock_msec() { return GetTickCount64(); }
 
 void az_platform_sleep_msec(int32_t milliseconds) { Sleep(milliseconds); }
 
-void az_platform_mtx_destroy(az_platform_mtx* mtx)
-{
-  DeleteCriticalSection(&mtx->_internal.cs);
-  *mtx = (az_platform_mtx){ 0 };
-}
-
-AZ_NODISCARD az_result az_platform_mtx_init(az_platform_mtx* mtx)
-{
-  InitializeCriticalSection(&mtx->_internal.cs);
-  return AZ_OK;
-}
-
-AZ_NODISCARD az_result az_platform_mtx_lock(az_platform_mtx* mtx)
-{
-  EnterCriticalSection(&mtx->_internal.cs);
-  return AZ_OK;
-}
-
-AZ_NODISCARD az_result az_platform_mtx_unlock(az_platform_mtx* mtx)
+AZ_NODISCARD bool az_platform_atomic_compare_exchange(
+    void* volatile* obj,
+    void* expected,
+    void* desired)
 {
-  LeaveCriticalSection(&mtx->_internal.cs);
-  return AZ_OK;
+  return InterlockedCompareExchangePointer(obj, expected, desired) == expected;
```

Comment:
@JeffreyRichter, I'll need to update this. https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchangepointer Says:
"On a 64-bit system, the parameters are 64 bits and must be aligned on 64-bit boundaries; otherwise, the function will behave unpredictably. On a 32-bit system, the parameters are 32 bits and must be aligned on 32-bit boundaries."
I think we need to store an array of 15 bytes, and then expect that whatever within this array ends up being aligned to a 64-bit boundary, that is our pointer. Because otherwise we need to require that the pointer is aligned, which means that the client needs to be aligned, which means a blob client, for example, can't be created on the stack.

Reply:
Don't worry about this; the compiler WILL ensure that the pointer is aligned even if allocated on the stack.

Reply:
Currently, Linux 64 and only Linux 64 unit tests hang. It is hard to tell what's going on, because preconditions just sleep forever, and I don't know which one failed. But the symptoms are similar. Or maybe on Linux 64 + GCC, compiler does not do this alighnment, and fails into UB, that's another possibility. However `__sync_bool_compare_and_swap` do not mention requirement for alignment.

Reply:
I have a suspicion that GCC on Linux64 does not align the pointer and ends up shooting into access token buffer memory.

Reply:
All compilers will align pointers. Plus, if a non-aligned value is accessed in a way that requires alignment, the system would not hang; it would generate an interrupt or exception instead. For example, on Windows, a EXCEPTION_DATATYPE_MISALIGNMENT exception is raised.

Reply:
To debug what's happening, run the app under the debugger and put a breakpoint on the failed preconditon function.  Or, override the precondition callback by calling https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/core/core/inc/az_precondition.h#L62 and have it to something else. For example, on Windows, you could call https://docs.microsoft.com/en-us/windows/win32/api/debugapi/nf-debugapi-debugbreak.  Or dump the stack somehow. It would be best to run in a debugger.

Reply:
> To debug what's happening, run the app under the debugger and put a breakpoint on the failed preconditon function.

FWIW, I couldn't get this approach to work in VS. Let me know if you are able to, @antkmsft

Reply:
NOTE: You have the arguments in the wrong order: 
PVOID InterlockedCompareExchangePointer(
  PVOID volatile *Destination,
  PVOID          Exchange,
  PVOID          Comperand
);

requires "expected" to be the LAST argument and "desired" is the 2nd argument.

Reply:
All is corrected, and works. I am resolving this conversation thread.